lib/Sema/CSSimplify.cpp

```diff
@@ -1543,7 +1546,8 @@ ConstraintSystem::matchTypes(Type type1, Type type2, ConstraintKind kind,
         // Simplify the right-hand type and perform the "occurs" check.
         typeVar1 = getRepresentative(typeVar1);
         type2 = simplifyType(type2, flags);
-        if (typeVarOccursInType(typeVar1, type2))
+        if (typeVarOccursInType(typeVar1, type2) ||
+            type2->is<DependentMemberType>())
```

Comment:
In the future could we make this check a separate closure e.g. `isBindable(TypeVariableType *, Type)` and use it in all 4 places?